src/single_random_draw.rs

```diff
@@ -0,0 +1,488 @@
+    input: &TxIn,
+    fee_rate: FeeRate,
+) -> Result<Option<Amount>, Error> {
+    let weight_predict = transaction::InputWeightPrediction::new(
+        input.script_sig.len(),
+        input.witness.iter().map(|elem| elem.len()),
+    );
+
+    let input_weight = transaction::predict_weight(Some(vec![weight_predict]), None::<Vec<_>>);
+    let input_fee: Option<Amount> = fee_rate.checked_mul_by_weight(input_weight);
+
+    match input_fee {
+        Some(f) => Ok(utxo.value.checked_sub(f)),
+        None => Err(Error::Multiplication(input_weight, fee_rate)),
+    }
+}
+
+/// Calculate the fee for a transaction with two UTXOs and no inputs.
+///
+/// Returns `Ok(Amount)` where Amount is the fee.
+///
+/// ### Errors
+///
+/// Returns `Err(Error::Multiplication)` if `FeeRate` * `Weight` overflows.
+///
+fn calculate_fee(
+    script_pubkey: &ScriptBuf,
+    fee_rate: FeeRate,
+    transaction: &mut Transaction,
+) -> Result<Amount, Error> {
+    let output: TxOut = TxOut { value: Amount::ZERO, script_pubkey: script_pubkey.clone() };
+
+    let output_two = output.clone();
+
+    transaction.output = vec![output, output_two];
+
+    let weight: Weight = transaction.weight();
+
+    let transaction_fee: Option<Amount> = fee_rate.checked_mul_by_weight(weight);
+
+    match transaction_fee {
+        Some(f) => Ok(f),
+        None => Err(Error::Multiplication(weight, fee_rate)),
+    }
+}
+
+/// Randomly select coins for the given target by shuffling the UTXO pool and
+/// taking UTXOs until the given target is reached.
+///
+/// The fee_rate can have an impact on the selection process since the fee
+/// must be paid for in addition to the target.  However, the total fee
+/// is dependant on the number of UTXOs consumed and the new inputs created.
+/// The selection strategy therefore calculates the fees of what is known
+/// ahead of time (the number of UTXOs create and the transaction header),
+/// and then then for each new input, the effective_value is tracked which
+/// deducts the fee for each individual input at selection time.  For more
+/// discussion see the following:
+///
+/// https://bitcoin.stackexchange.com/questions/103654/calculating-fee-based-on-fee-rate-for-bitcoin-transaction/114847#114847
+///
+/// ## Parameters
+/// ///
+/// /// * `target` - target value to send to recipient.
+/// /// * `fee_rate` - ratio of transaction amount per size.
+/// /// * `utxos` - UTXOs from which to sum the target amount.
+/// /// * `inputs` - the order corresponds to the ordering of the UTXOs where the first input consumes the first UTXO and so on.
+/// /// * `transaction` - the transaction to populate with resulting inputs and outputs.
+/// /// * `script_pubkey` - needed to compose the transaction outputs.
+/// /// * `rng` - used primarily by tests to make the selection deterministic.
+///
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
+    target: Amount,
+    fee_rate: FeeRate,
+    utxos: &mut [TxOut],
+    inputs: &mut [TxIn],
```

Comment:
You seem to be assuming that the `TxIn` already has the witness elements filled in, but that's usually not the case. I think we already had this discussion a while ago, but you probably need to either have the user provide the maximum witness length or use miniscript so that you can compute it.

Reply:
Is it unreasonable to assume that for each `UTXO` passed, that the wallet software using the interface would create a corresponding `TxIn`?  Maybe there is too much of a performance penalty?

> or use miniscript so that you can compute it.

Are you suggesting that miniscript can compute the size of the witness elements given a UTXO?  I need to look more into how miniscript works.  My understanding is just that it's a simplified version of bitcoin script.

Reply:
> Is it unreasonable to assume that for each UTXO passed, that the wallet software using the interface would create a corresponding TxIn? Maybe there is too much of a performance penalty?

It's not only a question of performance, but it's just not very practical. I personally wouldn't use a library that forces me to generate the witnesses for all of my UTXOs before they are even selected. I'd be fine with manually passing the maximum witness weight though.

> Are you suggesting that miniscript can compute the size of the witness elements given a UTXO? I need to look more into how miniscript works. My understanding is just that it's a simplified version of bitcoin script.

Yes miniscript can tell you what is the maximum witness length, see https://docs.rs/miniscript/latest/miniscript/descriptor/enum.Descriptor.html#method.max_weight_to_satisfy

Reply:
Besides, Witness, it looks like [ScriptBuf](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L187 ) is also an unknown size since it's a [vec](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/script/owned.rs#L35).  Maybe it's better to pass in the maximum input size?

Reply:
I'll need to do some more research here since there isn't much else for me to go on here.

Reply:
Ok, so it looks like for each TxOut, we compute the corresponding TxIn as:

```
input_weight = (32 + 4 + 4) * 4 + input.max_weight_to_satisfy();
```

At first glance however at miniscript, it seems that miniscript works on a transaction, not an individual input.   Would it work to pass in `max_satisfaction_weight` for the TxIn that corresponds to each UTXO?  That appears to be how the BDK implementation works: https://github.com/bitcoindevkit/bdk/blob/feafaaca31a0a40afc03ce98591d151c48c74fa2/crates/bdk/src/wallet/coin_selection.rs#L716

Reply:
> At first glance however at miniscript, it seems that miniscript works on a transaction, not an individual input.

No you can generate descriptors that represent spending conditions for a single output.

> Would it work to pass in max_satisfaction_weight for the TxIn that corresponds to each UTXO?

Yes, that's what I was referring to when I wrote:
> I'd be fine with manually passing the maximum witness weight though.

Reply:
I see.  There really isn't very much when I search for "satisfaction weight".  So, the satisfaction weight is the base input weight plus the size of the spending conditions of the corresponding UTXO.  I'll take a closer look at miniscript because it would be nice if the selection algo could compute the "satisfaction weight" instead of needing to be passed as a param.

Reply:
AFAIK satisfaction weight is the maximum weight of the parameters that need to be given to satisfy the spending condition, see e.g. how they compute the `P2WPKH_SATISFACTION_SIZE_CONSTANT`: https://github.com/bitcoindevkit/bdk/blob/master/crates/bdk/src/wallet/coin_selection.rs#L704

And tbh if I were you I'd go first with something like that, get a something usable and then maybe later miniscript if it feels like it would add value.

Reply:
> AFAIK satisfaction weight is the maximum weight of the parameters that need to be given to satisfy the spending condition, see e.g. how they compute the

Yeah, after further inspection of the bitcoin source [here](https://github.com/bitcoin/bitcoin/blob/master/src/script/miniscript.h#L38) it looks like satisfaction is used to describe an argument passed to the stack that "satisfies" the requirement.  So the satisfaction weight is the size of all the required params that get pushed on the stack.